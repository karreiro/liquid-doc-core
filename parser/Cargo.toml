[package]
name = "liquid_doc_parser"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["rlib"]

[dependencies]
serde.workspace = true
serde_json.workspace = true
pest = "2.7"
pest_derive = "2.7"

[dev-dependencies]
insta.workspace = true
paste = "1.0"
pretty_assertions.workspace = true
