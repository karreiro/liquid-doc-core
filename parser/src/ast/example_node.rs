use serde::{Deserialize, Serialize};

use super::{position::Position, LiquidNode, TextNode};

const NODE_NAME: &str = "example";
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct LiquidDocExampleNode {
    pub name: String,
    pub position: Position,
    pub source: String,
    pub content: Box<LiquidNode>,
    #[serde(rename = "isInline")]
    pub is_inline: bool,
}

impl LiquidDocExampleNode {
    pub fn new(content: TextNode, is_inline: bool, position: Position, source: String) -> Self {
        LiquidDocExampleNode {
            content: Box::new(LiquidNode::TextNode(content)),
            is_inline,
            position,
            source,
            name: NODE_NAME.to_string(),
        }
    }

    pub fn from_pair(
        pair: &pest::iterators::Pair<crate::parser::Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        assert!(
            pair.as_rule() == crate::parser::Rule::ExampleNode,
            "Expected a exampleNode, found {:?}",
            pair.as_rule()
        );

        let mut content = TextNode::from_pair(pair, position_offset);
        content.trim_content_start("@example ");

        let position = Position::from_pair(pair, position_offset);
        let source = pair.as_str().to_string();
        LiquidDocExampleNode::new(content, true, position, source)
    }
}

#[cfg(test)]
mod test {
    use crate::{assert_json_output, ast::LiquidNode, parser::parse_liquid_string};

    use pretty_assertions::assert_eq;

    #[test]
    fn parse_inline_example_node() {
        let input = "@example simple inline example\n";
        let result = parse_liquid_string(input, Some(10));

        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocExampleNode(example_node) = node {
            assert_eq!(
                example_node.content.as_text_node_unsafe().as_str(),
                "simple inline example\n"
            );
            assert!(example_node.is_inline);
        } else {
            panic!("Expected a LiquidDocExampleNode");
        }
    }

    #[test]
    pub fn test_serialization_round_trip() {
        assert_json_output!("@example simple inline example\n");
    }

    #[test]
    pub fn complex_example() {
        assert_json_output!("@example
{% render 'resource-card', resource: product, resource_type: 'product', image_width: 300, image_aspect_ratio: '1/1' %}
");
    }
}
