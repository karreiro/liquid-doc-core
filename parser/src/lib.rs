mod ast;
mod parser;

// Re-export all AST types
pub use ast::*;

// Re-export parser functionality
pub use parser::{parse_liquid_string, visit, LiquidParser};
