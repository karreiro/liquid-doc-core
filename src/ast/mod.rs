mod description_node;
mod doc_param_node;
mod example_node;
mod liquid_ast;
mod position;
pub(crate) mod test_utils;
mod text_node;

pub use description_node::*;
pub use doc_param_node::*;
pub use example_node::*;
pub use liquid_ast::*;
pub use text_node::*;
