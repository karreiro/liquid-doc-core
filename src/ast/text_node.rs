use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::position::Position;

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct TextNode {
    pub value: String,
    pub position: Position,
    pub source: String,
    #[serde(rename = "type", skip_deserializing)]
    pub type_: &'static str,
}

impl TextNode {
    pub fn new(value: String, position: Position, source: String) -> Self {
        TextNode {
            value,
            position,
            source,
            type_: "TextNode",
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, position_offset: Option<usize>) -> Self {
        let source_str = pair.as_str();
        TextNode::new(
            source_str.to_string(),
            Position::from_pair(pair, position_offset),
            source_str.to_string(),
        )
    }

    pub fn without_brackets(
        pair: &pest::iterators::Pair<Rule>,
        position_offset: Option<usize>,
    ) -> Self {
        let mut text_node = Self::from_pair(pair, position_offset);
        let new_value = text_node
            .value
            .trim_matches(|c| c == '{' || c == '}' || c == '[' || c == ']')
            .to_string();
        if new_value == text_node.value {
            return text_node; // No change needed
        }
        text_node.value = new_value;

        text_node.position.shift_start(1); // Adjust position to account for removed brackets
        text_node.position.shift_end_down(1); // Adjust end position as well
        text_node
    }

    pub fn trim_content_start(&mut self, to_strip: &str) {
        if self.value.starts_with(to_strip) {
            self.value = self.value.trim_start_matches(to_strip).to_string();
            self.position.shift_start(to_strip.len());
        }
    }

    #[allow(dead_code)]
    pub fn as_str(&self) -> &str {
        &self.value
    }
    pub fn is_empty(&self) -> bool {
        self.value.is_empty()
    }
}
