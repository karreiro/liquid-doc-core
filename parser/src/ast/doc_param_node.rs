use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::position::Position;
use super::text_node::TextNode;
use super::LiquidNode;

const NODE_NAME: &str = "param";
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct LiquidDocParamNode {
    pub name: String,
    pub position: Position,
    pub source: String,
    #[serde(rename = "paramName")]
    pub param_name: Box<LiquidNode>,
    #[serde(rename = "paramDescription")]
    pub param_description: Option<Box<LiquidNode>>,
    #[serde(rename = "paramType")]
    pub param_type: Option<Box<LiquidNode>>,
    pub required: bool,
}
impl LiquidDocParamNode {
    fn new(
        position: Position,
        source: String,
        param_type: Option<TextNode>,
        param_name: TextNode,
        param_description: Option<TextNode>,
        required: bool,
    ) -> Self {
        LiquidDocParamNode {
            name: NODE_NAME.to_string(),
            position,
            source,
            param_type: param_type.map(|t| Box::new(LiquidNode::TextNode(t))),
            param_name: Box::new(LiquidNode::TextNode(param_name)),
            param_description: param_description.map(|d| Box::new(LiquidNode::TextNode(d))),
            required,
        }
    }
    pub fn from_pair(pair: &pest::iterators::Pair<Rule>, position_offset: Option<usize>) -> Self {
        assert!(
            pair.as_rule() == Rule::ParamNode,
            "Expected a paramNode, found {:?}",
            pair.as_rule()
        );

        let mut inner = pair.clone().into_inner();

        let first = inner.next().expect("Expected at least one inner pair");

        let (param_type, name) = if let Rule::paramType = first.as_rule() {
            (
                Some(TextNode::without_brackets(&first, position_offset)),
                inner.next().expect("Expected a paramName after paramType"),
            )
        } else {
            (None, first)
        };
        let original_name_str = name.as_str();
        let required = !original_name_str.starts_with('[') && !original_name_str.ends_with(']');

        let param_name = TextNode::without_brackets(&name, position_offset);

        let description = inner.next().and_then(|t| {
            if !t.as_str().is_empty() {
                Some(TextNode::from_pair(&t, position_offset))
            } else {
                None
            }
        });

        LiquidDocParamNode::new(
            Position::from_pair(pair, position_offset),
            pair.as_str().to_string(),
            param_type,
            param_name,
            description,
            required,
        )
    }
}

#[cfg(test)]
mod tests {
    use crate::assert_json_output;
    use crate::ast::LiquidNode;
    use crate::parser::parse_liquid_string;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_parse_param_with_type() {
        let input = "@param {sometype} requiredParamWithNoType";
        let result = parse_liquid_string(input, Some(10));

        assert!(result.is_some());
        let node = result.unwrap().head();

        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.as_text_node_unsafe().as_str(),
                "requiredParamWithNoType"
            );
            assert!(param_node.param_description.is_none());
            assert!(param_node.param_type.is_some());
            assert_eq!(
                param_node
                    .param_type
                    .unwrap()
                    .as_text_node_unsafe()
                    .as_str(),
                "sometype"
            );
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }
    #[test]
    fn test_parse_param_with_type_and_description() {
        let input = "@param {sometype} requiredParamWithNoType - This is a cool parameter";
        let result = parse_liquid_string(input, Some(10));

        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.as_text_node_unsafe().as_str(),
                "requiredParamWithNoType"
            );
            assert_eq!(
                param_node
                    .param_description
                    .unwrap()
                    .as_text_node_unsafe()
                    .as_str(),
                "This is a cool parameter"
            );
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }

    #[test]
    fn test_parse_optional_param_with_type_and_description() {
        let input =
            "@param {sometype} [optionalParamWithTypeAndDescription] - This is a cool parameter";
        let result = parse_liquid_string(input, Some(10));

        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.as_text_node_unsafe().as_str(),
                "optionalParamWithTypeAndDescription"
            );
            assert_eq!(
                param_node
                    .param_description
                    .unwrap()
                    .as_text_node_unsafe()
                    .as_str(),
                "This is a cool parameter"
            );
            assert!(!param_node.required)
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }

    #[test]
    pub fn test_serialization_round_trip() {
        assert_json_output!(
            "@param {sometype} requiredParamWithSomeType - This is a cool parameter"
        );
    }

    #[test]
    pub fn complex_example_with_many_params() {
        assert_json_output!("@param requiredParamWithNoType
@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.
@param {String} paramWithNoDescription
@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description
@param [optionalParameterWithDescription] - optional parameter description
@param {String} [optionalParameterWithType]")
    }
}
