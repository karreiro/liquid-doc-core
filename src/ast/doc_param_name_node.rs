use serde::{Deserialize, Serialize};

use crate::{ast::position::Position, parser::Rule};

use super::text_node::TextNode;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LiquidDocParamNameNode {
    pub content: TextNode,
    pub position: Position,
    pub source: String,
    pub required: bool,
}

impl LiquidDocParamNameNode {
    pub fn new(pair: &pest::iterators::Pair<Rule>) -> Self {
        assert!(
            pair.as_rule() == Rule::paramName,
            "Expected a paramName, found {:?}",
            pair.as_rule()
        );

        let source_str = pair.as_str();

        LiquidDocParamNameNode {
            content: TextNode::new(pair),
            position: Position::from_pair(pair),
            source: source_str.to_string(),
            required: !source_str.starts_with('[') && !source_str.ends_with(']'),
        }
    }
}
