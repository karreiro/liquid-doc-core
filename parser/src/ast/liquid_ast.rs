use crate::ast::{LiquidDocDescriptionNode, LiquidDocParamNode, TextNode};
use serde::{Deserialize, Serialize};

use super::LiquidDocExampleNode;

/// Represents the different types of nodes in a Liquid AST.
/// Each variant corresponds to a specific type of node in the Liquid template language.
/// If a node has a specific content type, it is represented as a `LiquidNode` variant, rather than the more specific type.
/// This instructs the serializer to use the `type` field to differentiate between node types, which the consuming code requires.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
#[serde(tag = "type")]
#[allow(clippy::enum_variant_names)]
pub enum LiquidNode {
    LiquidDocDescriptionNode(LiquidDocDescriptionNode),
    TextNode(TextNode),
    LiquidDocParamNode(LiquidDocParamNode),
    LiquidDocExampleNode(LiquidDocExampleNode),
}

impl LiquidNode {
    #[cfg(test)]
    pub fn as_text_node_unsafe(&self) -> &TextNode {
        if let LiquidNode::TextNode(text_node) = self {
            text_node
        } else {
            panic!("Tried to access TextNode from a non-TextNode LiquidNode");
        }
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub struct LiquidAST {
    pub nodes: Vec<LiquidNode>,
}

impl LiquidAST {
    pub fn new() -> Self {
        LiquidAST { nodes: Vec::new() }
    }

    pub fn add_node(&mut self, node: LiquidNode) {
        self.nodes.push(node);
    }

    #[allow(dead_code)]
    pub fn head(&self) -> LiquidNode {
        self.nodes[0].clone()
    }
}
