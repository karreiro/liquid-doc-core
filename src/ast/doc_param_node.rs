use serde::{Deserialize, Serialize};

use crate::parser::Rule;

use super::doc_param_name_node::LiquidDocParamNameNode;
use super::text_node::TextNode;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct LiquidDocParamNode {
    pub name: String,
    #[serde(rename = "locStart")]
    pub loc_start: usize,
    #[serde(rename = "locEnd")]
    pub loc_end: usize,
    pub source: String,
    #[serde(rename = "paramType")]
    pub param_type: Option<TextNode>,
    #[serde(rename = "paramName")]
    pub param_name: LiquidDocParamNameNode,
    #[serde(rename = "paramDescription")]
    pub param_description: Option<TextNode>,
}
impl LiquidDocParamNode {
    pub fn new(pair: &pest::iterators::Pair<Rule>) -> Self {
        println!("{:#?}", pair);
        assert!(
            pair.as_rule() == Rule::paramNode,
            "Expected a paramNode, found {:?}",
            pair.as_rule()
        );

        let mut inner = pair.clone().into_inner();

        let first = inner.next().expect("Expected at least one inner pair");

        let (param_type, name) = if let Rule::paramType = first.as_rule() {
            (
                // Remove the curly braces from the type string
                Some(TextNode::new_trim_ends(&first)),
                inner.next().expect("Expected a paramName after paramType"),
            )
        } else {
            (None, first)
        };

        let param_name = LiquidDocParamNameNode::new(&name);

        let description = inner.next().and_then(|t| {
            if !t.as_str().is_empty() {
                Some(TextNode::new(&t))
            } else {
                None
            }
        });

        // let pair = pair.next
        let source_str = pair.as_str();
        let span = pair.as_span();
        LiquidDocParamNode {
            name: "param".to_string(), // The node name is always "param"
            loc_start: span.start(),
            loc_end: span.end(),
            source: source_str.to_string(),
            param_type, // Default to None, can be set later
            param_name,
            param_description: description,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::ast::LiquidNode;
    use crate::parser::parse_liquid_string;

    /*
        """@param requiredParamWithNoType
            @param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.
            @param {String} paramWithNoDescription
            @param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description
            @param [optionalParameterWithDescription] - optional parameter description
            @param {String} [optionalParameterWithType]
            @unsupported this node falls back to a text node
    """
         */
    #[test]
    fn test_parse_param_with_type() {
        let input = "@param {sometype} requiredParamWithNoType";
        let result = parse_liquid_string(input);

        println!("{:#?}", result);
        assert!(result.is_some());
        let node = result.unwrap().head();

        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.content.as_str(),
                "requiredParamWithNoType"
            );
            assert!(param_node.param_description.is_none());
            assert!(param_node.param_type.is_some());
            assert_eq!(param_node.param_type.unwrap().as_str(), "sometype");
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }
    #[test]
    fn test_parse_param_with_type_and_description() {
        let input = "@param {sometype} requiredParamWithNoType - This is a cool parameter";
        let result = parse_liquid_string(input);

        println!("{:#?}", result);
        assert!(result.is_some());
        let node = result.unwrap().head();
        if let LiquidNode::LiquidDocParamNode(param_node) = node {
            assert_eq!(
                param_node.param_name.content.as_str(),
                "requiredParamWithNoType"
            );
            assert_eq!(
                param_node.param_description.unwrap().as_str(),
                "This is a cool parameter"
            );
        } else {
            panic!("Expected a LiquidDocParamNode");
        }
    }
}
