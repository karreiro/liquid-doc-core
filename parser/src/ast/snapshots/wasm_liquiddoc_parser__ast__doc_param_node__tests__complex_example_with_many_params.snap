---
source: src/ast/doc_param_node.rs
expression: "@param requiredParamWithNoType\n@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.\n@param {String} paramWithNoDescription\n@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description\n@param [optionalParameterWithDescription] - optional parameter description\n@param {String} [optionalParameterWithType]"
---
[
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 10,
      "end": 40
    },
    "source": "@param requiredParamWithNoType",
    "paramName": {
      "type": "TextNode",
      "value": "requiredParamWithNoType",
      "position": {
        "start": 17,
        "end": 40
      },
      "source": "requiredParamWithNoType"
    },
    "paramDescription": null,
    "paramType": null,
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 41,
      "end": 161
    },
    "source": "@param {String} paramWithDescription - param with description and `punctation`. This is still a valid param description.",
    "paramName": {
      "type": "TextNode",
      "value": "paramWithDescription",
      "position": {
        "start": 57,
        "end": 77
      },
      "source": "paramWithDescription"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "param with description and `punctation`. This is still a valid param description.",
      "position": {
        "start": 80,
        "end": 161
      },
      "source": "param with description and `punctation`. This is still a valid param description."
    },
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 49,
        "end": 55
      },
      "source": "{String}"
    },
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 162,
      "end": 200
    },
    "source": "@param {String} paramWithNoDescription",
    "paramName": {
      "type": "TextNode",
      "value": "paramWithNoDescription",
      "position": {
        "start": 178,
        "end": 200
      },
      "source": "paramWithNoDescription"
    },
    "paramDescription": null,
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 170,
        "end": 176
      },
      "source": "{String}"
    },
    "required": true
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 201,
      "end": 305
    },
    "source": "@param {String} [optionalParameterWithTypeAndDescription] - optional parameter with type and description",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithTypeAndDescription",
      "position": {
        "start": 218,
        "end": 257
      },
      "source": "[optionalParameterWithTypeAndDescription]"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "optional parameter with type and description",
      "position": {
        "start": 261,
        "end": 305
      },
      "source": "optional parameter with type and description"
    },
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 209,
        "end": 215
      },
      "source": "{String}"
    },
    "required": false
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 306,
      "end": 380
    },
    "source": "@param [optionalParameterWithDescription] - optional parameter description",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithDescription",
      "position": {
        "start": 314,
        "end": 346
      },
      "source": "[optionalParameterWithDescription]"
    },
    "paramDescription": {
      "type": "TextNode",
      "value": "optional parameter description",
      "position": {
        "start": 350,
        "end": 380
      },
      "source": "optional parameter description"
    },
    "paramType": null,
    "required": false
  },
  {
    "type": "LiquidDocParamNode",
    "name": "param",
    "position": {
      "start": 381,
      "end": 424
    },
    "source": "@param {String} [optionalParameterWithType]",
    "paramName": {
      "type": "TextNode",
      "value": "optionalParameterWithType",
      "position": {
        "start": 398,
        "end": 423
      },
      "source": "[optionalParameterWithType]"
    },
    "paramDescription": null,
    "paramType": {
      "type": "TextNode",
      "value": "String",
      "position": {
        "start": 389,
        "end": 395
      },
      "source": "{String}"
    },
    "required": false
  }
]
